cupyx/scipy/stats/_stats_py.py

```diff
@@ -0,0 +1,147 @@
+    with length 1, so it can be used in cupy.apply_along_axis.
+
+    """
+
+    y = x[~cupy.isnan(x)]
+    if y.size == 0:
+        return cupy.array([True])
+    else:
+        return (y[0] == y).all(keepdims=True)
+
+
+def _quiet_nanmean(x):
+    """Compute nanmean for the 1d array x, but quietly return
+    nan if x is all nan. The return value is a 1d array with
+    length 1, so it can be used in cupy.apply_along_axis.
+
+    """
+
+    y = x[~cupy.isnan(x)]
+    if y.size == 0:
+        return cupy.array([cupy.nan])
+    else:
+        return cupy.mean(y, keepdims=True)
+
+
+def _quiet_nanstd(x, ddof=0):
+    """Compute nanstd for the 1d array x, but quietly return
+    nan if x is all nan. The return value is a 1d array
+    with length 1, so it can be used in cupy.apply_along_axis
+
+    """
+
+    y = x[~cupy.isnan(x)]
+    if y.size == 0:
+        return cupy.array([cupy.nan])
+    else:
+        return cupy.std(y, keepdims=True, ddof=ddof)
+
+
+def zmap(scores, compare, axis=0, ddof=0, nan_policy='propagate'):
+    """Calculate the relative z-scores.
+
+    Return an array of z-scores, i.e., scores that are standardized
+    to zero mean and unit variance, where mean and variance are
+    calculated from the comparison array.
+
+    Parameters
+    ----------
+    scores : array-like
+        The input for which z-scores are calculated
+    compare : array-like
+        The input from which the mean and standard deviation of
+        the normalization are taken; assumed to have the same
+        dimension as `scores`
+    axis : int or None, optional
+        Axis over which mean and variance of `compare` are calculated.
+        Default is 0. If None, compute over the whole array `scores`
+    ddof : int, optional
+        Degrees of freedom correction in the calculation of the
+        standard deviation. Default is 0
+    nan_policy : {'propagate', 'raise', 'omit'}, optional
+        Defines how to handle the occurrence of nans in `compare`.
+        'propagate' returns nan, 'raise' raises an exception, 'omit'
+        performs the calculations ignoring nan values. Default is
+        'propagate'. Note that when the value is 'omit', nans in `scores`
+        also propagate to the output, but they do not affect the z-scores
+        computed for the non-nan values
+
+    Returns
+    -------
+    zscore : array-like
+        Z-scores, in the same shape as `scores`
+
+    """
+
+    a = compare
+
+    if a.size == 0:
+        return cupy.empty(a.shape)
+
+    contains_nan, nan_policy = _contains_nan(a, nan_policy)
+
+    if contains_nan and nan_policy == 'omit':
+        if axis is None:
+            mn = _quiet_nanmean(a.ravel())
+            std = _quiet_nanstd(a.ravel(), ddof=ddof)
+            isconst = _isconst(a.ravel())
+        else:
+            mn = cupy.apply_along_axis(_quiet_nanmean, axis, a)
+            std = cupy.apply_along_axis(_quiet_nanstd, axis, a, ddof=ddof)
```

Comment:
CuPy has `cupy.apply_along_axis`, but it is quite inefficient because it repeatedly launches kernels for 1d-arrays along the axis. We may simply use `cupy.nanmean` and `cupy.nanstd`.

Reply:
Hi, @takagi!
I've removed `apply_along_axis` and used `cupy.nanmean` and `cupy.nanstd`. But I see there is one drawback to this:
- `cupy.nanstd` that uses `cupy.nanvar` which is not implemented for complex dtypes. But we will require cases where we use nan_polity=`omit` and require the `axis` argument.

Using apply_along_axis resolves these problems because we use `mean` and `std` that support complex dtypes.

Anyways, for now, I've removed the `apply_along_axis`, considering performance as the first priority.

What would you prefer? Can we add support for complex dtypes in `nanvar`?

Thank you!

Reply:
Let's support complex numbers in `cupy.nanvar`. I'll put a pointer soon.

Reply:
When we remove the assertion at 
https://github.com/cupy/cupy/blob/fc04c9445f488272c9a1b78999e4e0df15a721b5/cupy/_core/_routines_statistics.pyx#L698

, we'll get the following error:

```
TypeError: Wrong type ((<class 'numpy.complex64'>,)) of arguments for cupy_count_non_nan
```

Let's add complex support to `_count_non_nan`, that is, `F->q` and `D->q`

https://github.com/cupy/cupy/blob/fc04c9445f488272c9a1b78999e4e0df15a721b5/cupy/_core/_routines_statistics.pyx#L682-L685

, then we'll get the following compile error:

```
cupy.cuda.compiler.CompileException: /tmp/tmp380imuka/328e8b517679a3804b06c7d50f2bc4fa55c4a2df.cubin.cu(42): error: no operator "/" matches
these operands
            operand types are: const T / const long long

/tmp/tmp380imuka/328e8b517679a3804b06c7d50f2bc4fa55c4a2df.cubin.cu(11): error: no operator "/" matches these operands
            operand types are: T / long long
          detected during instantiation of "T nanvar_impl(S, T, long long) [with S=S, T=T]"
(42): here
```

The related complex division is defined in 

https://github.com/cupy/cupy/blob/fc04c9445f488272c9a1b78999e4e0df15a721b5/cupy/_core/include/cupy/complex/arithmetic.h#L110-L114

, where it requires that the divisor has the same type as the dividend's element type `T`. I think we can relax the constraint with another typename `S`.

Would you make another PR to support complex numbers in `cupy.nanvar`?

Reply:
Interesting, Thanks for the clear description!

> Would you make another PR to support complex numbers in cupy.nanvar?

Definitely, I'll send a PR within some hours!

Reply:
Thanks!